sim/dex-data.ts

```diff
@@ -34,7 +34,21 @@ export function toID(text: any): ID {
 	/* eslint-enable @typescript-eslint/prefer-optional-chain */
 }
 
-export class BasicEffect implements EffectData {
+/**
+ * Like Object.assign but only assigns fields missing from self.
+ * Facilitates consistent field ordering in constructors.
+ * Modifies self in-place.
+ */
+export function assignNewFields(self: AnyObject, data: AnyObject) {
+	for (const k in data) {
+		// to behave like Object.assign, skip data's inherited properties
+		if (!Object.prototype.hasOwnProperty.call(data, k)) continue;
+		if (Object.prototype.hasOwnProperty.call(self, k)) continue;
+		self[k] = data[k];
+	}
+}
```

Comment:
I'd prefer for this to be a method on BasicEffect, assuming there are no performance impacts there.

Reply:
I don't think there's any perf problems with that, but I do think it would be counter-intuitive to have a method that can only be called during construction. I see this more as a utility method than something specific to `BasicEffect`.

Reply:
I guess we rarely export bare functions outside of `lib/`. My mind is saying either move to `lib/utils` or make it a static method on `BasicEffect`, either would make it feel more right, but yeah it's not really a big deal.

Reply:
I'm happy to follow your lead here - I'm pretty much a blank slate on Typescript code style.

Reply:
Hm. On further reflection, this is probably fine this way.